doc/developer/design/20240610_unified_compute_introspection.md

```diff
@@ -0,0 +1,134 @@
+# Unified Compute Introspection
+
+- Associated: 
+  - https://github.com/MaterializeInc/materialize/issues/26730
+  - https://github.com/MaterializeInc/materialize/issues/26682
+  - https://github.com/MaterializeInc/materialize/issues/21835
+
+## The Problem
+
+Today compute has an infrastructure of per-replica logging dataflows that collect various kinds of introspection information (frontiers, scheduling durations, arrangement sizes, ...) and expose them in arrangements.
+The contents of these arrangements naturally differ between replicas, so querying them requires the use of replica-targeted queries.
+This introduces two major pain points:
+
+- Collecting introspection information for all replicas requires first enumerating the live replicas, then sending a replica-targeted query for each of them, then merging the received results.
+  In a lot of contexts (e.g. monitoring tools) this workflow is hard or impossible to implement.
+- If a replica is unresponsive, targeted queries against it will hang and/or timeout, removing the ability to collect any, even stale, introspection data.
+
+These issues limit the usability of compute introspection for, e.g., use in the console and diagnosis of performance issues.
+
+## Success Criteria
+
+Compute provides a way to expose a subset of replica introspection data that:
+
+- ... can be retrieved through a single query whose definition is independent of the specifics of the exporting replicas.
+- ... can be retrieved successfully regardless of the health of any of the exporting replicas.
+
+## Out of Scope
+
+- Providing freshness of unified introspection data.
+  If a replica is unresponsive, its retrievable introspection data might be stale.
+
+## Solution Proposal
+
+This design describes an MVP implementation for Unified Compute Introspection that can provide value quickly.
+
+### Overview
+
+- The compute controller defines a set of _introspection subscribes_ that continuously read from introspection indexes, optionally transforming (filter/aggregate/join/...) the read data.
+- The catalog defines a set of `BuiltinSource`s and associated `IntrospectionType`s, one for each introspection subscribe.
+- Whenever a replica connection is established, the compute controller installs all defined introspection subscribes on the replica, as replica-targeted subscribes.
+- Whenever a batch of updates arrives for an introspection subscribe, the compute controller (a) prepends the replica ID to all updates and (b) sends the thus enriched updates to the storage controller, tagged with the corresponding `IntrospectionType`.
+- The storage controller's `CollectionManager` appends the introspection updates to the collection associated with the given `IntrospectionType`.
+- Whenever a replica disconnects, the compute controller cancels its introspection subscribes and instructs the storage controller to delete all introspection data previously emitted for this replica.
```

Comment:
Just to be clear, this means it would emit retractions for the data, so a `SUBSCRIBE AS OF` would still include data for the dropped replica, correct?

Reply:
Yes! In this regard the unified introspection relations behave like any system relation that contains a `replica_id` column (`mz_cluster_replicas`, `mz_cluster_replica_statuses`, ...).